### ISSUE TITLE
Do we have plan to offer C++ binding for prune related features. 

### ISSUE THREAD

**Issue by linjiapro** [pytorch/pytorch#45855]
## 🚀 Feature
On the python side, there is a range of classes/functions to support the pruning functionality of the neural network:
https://pytorch.org/docs/stable/search.html?q=prune&check_keywords=yes&area=default

The motivation to add pruning is mentioned well in this feature request when pruning is added on python side:
https://github.com/pytorch/pytorch/issues/20402

CC the folks who worked on the python side for this: @mickypaganini @ezyang

**Comment by linjiapro** [comment]
This makes me think about the interesting issue that usually a feature is implemented in python first, then migrate to C++, and other bindings. That means the same logic could get implemented multiple times. One way to resolve this is that a very low-level language is chosen to be the main one. The logic and features are implemented in the low-level language and then implement bindings for high-level languages. The advantage of this is that since most logic executions will happen in a low-level language, the speed will be very fast. More importantly, the same feature will not be implemented multiple times. For pytorch, it seems one way to go is to always implement logic/features in C++, and then offer python API.

**Comment by ezyang** [comment]
One reason we still like to have Python only implementation is because it makes PyTorch more hackable: let's say that PyTorch has an optimizer written in Python and you are a researcher and you need mostly the same thing, but maybe with some slight tweak. If the optimizer is in Python, you can just copy paste the code into your project and edit it as you need it. If it's in C++ this is a lot harder.
